copy.go

```diff
@@ -297,20 +314,24 @@ func StructToMap(filter FieldFilter, src interface{}, dst map[string]interface{}
 			} else {
 				newValue = make(map[string]interface{})
 			}
-			if err := StructToMap(subFilter, srcField.Interface(), newValue, userOpts...); err != nil {
+			if err := structToMap(subFilter, srcField.Interface(), newValue, userOptions); err != nil {
 				return err
 			}
 			dst[dstName] = newValue
 
 		case reflect.Array, reflect.Slice:
 			// Check if it is a slice of primitive values.
 			itemKind := srcField.Type().Elem().Kind()
+			srcLen := userOptions.CopyListSize(&srcField, nil)
```

Comment:
Hm... passing "nil" as a second parameter does not seem right.
We can either:

1. change the signature of the function to only accept one parameter (src)
2. or create 2 different functions (each for structToStruct and structToMap) with the following signatures: `CopyListSize func(src, dst *reflect.Value) int` and `CopyListSize func(src *reflect.Value, dst map[string]interface{}) int`. This approach will also require having 2 different options structs for structToStruct and structToMap.

I vote for option 1: `CopyListSize func(src *reflect.Value) int`

Reply:
I also agree with option 1, but There is a problem in optiotn 1： `CopyListSize func(src *reflect.Value) int`, which can't control `srcLen` through `FiledName`. Could We consider that let the visitor function following signatures :`CopyListSize(srcFiled reflect. StructField)`. This approach get the field names in the struct. However, this will bring about a big change to `StructToStruct`.

Reply:
It should be possible to create a pointer to the needed field and then use it in the CopyListSize function:

```go
type A struct {
	Field1 []int
	Field2 []int
}

src := A{Field1: []int{1, 2, 3}, Field2: []int{4, 5, 6}}

field2 := reflect.ValueOf(src).FieldByName("Field2")
dst := &A{}
mask := fieldmask_utils.MaskFromString("Field1, Field2")
err := fieldmask_utils.StructToStruct(mask, src, dst,
		fieldmask_utils.WithCopyListSize(func(src *reflect.Value) int {
                        if(src == &field2) {
                          return 2
                        }
			return copySize
		}))
```

Reply:
I don't think this way (i.e. creating a pointer to the needed field)  is good way, because of `Field1` and `Field2` can refer same slice. For example:
```golang
type A struct {
	Field1 []int
	Field2 []int
}

var arr = []int{1,2,3}

var src = &A{
	Field1: arr,
	Field2: arr,
}
var field1 = reflect.ValueOf(src).Elem().FieldByName("Field1")
var dst = &A{}
var mask = fieldmask_utils.MaskFromString("Field1,Field2")
var err = fieldmask_utils.StructToStruct(mask, src, dst, fieldmask_utils.WithCopyListSize(
	func(src *reflect.Value) int {
		if src.Pointer() == (&field1).Pointer() {
			return 2
		} else {
			return src.Len()
		}
	},
))
require.NoError(t, err)
assert.Equal(t, &A{
	Field1: []int{1, 2},
	Field2: []int{1, 2, 3},
}, dst)
```
I think you can seriously consider my option:  `CopyListSize(srcFiled reflect.StructField)`.

Reply:
Is there a practical use-case of having 2 different fields in a struct pointing exactly to the same slice/array? If there is such a use-case (which I can't come up with) then let's consider `CopyListSize(srcFiled reflect.StructField)`, although I'm curios how much refactoring it may require.

Reply:
I also tried to refactor `structToStruct` just now. I've found it almost impossible to refactor `structToStruct` with `CopyListSize(src *reflect.StructField)`. I'm also give up mine plan: `CopyListSize(src *reflect.StructField)`, in turn I vote your solution: `CopyListSize(src *reflect.Value)`. However does it need to add this limitation (i.e. can't set copy list size according to fieldname with same addr) in READMe.md.